### ISSUE TITLE
Apps can bypass VPN with killswitch by sending multicast packets

### ISSUE THREAD

**Issue by ryrona2** [GrapheneOS/os-issue-tracker#3443]
I have been trying out the security of the VPN functionality in GrapheneOS to see that there are no leaks or VPN bypass when killswitch is enabled. I have monitored the network traffic using Wireshark on my Wifi hotspot.

Here I have observed that apps are able to send multicast packets not going out over the VPN, but instead going to the local network. This even with apps like Mullvad VPN app that clearly says it blocks all local network traffic. This is a clear privacy issue in the common case for using a VPN where the local network is untrusted, because the user expects no one on the network learning about your apps or usage when on the VPN. This issue is reproducible with both Mullvad VPN app and official Wireguard app, and probably others, so likely an issue in GrapheneOS rather than the apps.

Steps to reproduce:
1. Install Mullvad VPN app from F-Droid or official Wireguard app from their website.
2. Login to or configure your VPN app, and make sure always-on and killswitch is enabled for the VPN in settings.
3. Install VLC app from F-Droid.
4. Start VLC and click to menus to browse videos, audio and other media. Broadcast packets are being sent out on the local network to scan for media sources, bypassing the VPN.

The broadcast packets are regular IP broadcast packets of UDP and IGMP kind. I don't know if the app can receive answers to these broadcasts.

**Comment by ryrona2** [comment]
Possible relevant forum thread, Spotify may be sending multicast packets as well:
https://discuss.grapheneos.org/d/10337-spotify-communicating-with-other-devices-on-local-network/6

**Comment by flawedworld** [comment]
This can be easily reproduced visually using https://play.google.com/store/apps/details?id=org.xbmc.kore and by exposing Kodi's management port as described in the app.

**Comment by ryrona2** [comment]
@maade93791 Yes. I have gone down that path already, and then to netd, and there I saw that netd sets up the kill switch using netlink to manipulate the routing table rules, and nothing else. So, I do not believe iptables is being used either.

@thestinger Really? I felt pretty confident I have found the actual logic being used, and that was manipulation of routing table rules. But I guess I will know soon anyway, I have made a userdebug build and am trying to figure out how one actually flashes the thing. After that I should be easily able to confirm whether I have found the right place or not by simply manually removing the routing table rules I believe is the killswitch and see what happens.

@no-usernames-left That could be so. I know the packets are definitely sent out, but not more than that. And since it isn't iptables being used, it is a bit hard for me to understand what the rules actually do block and not. But I will see if I can figure it out. It is a bit odd they didn't go with iptables, since iptables is designed to support use-cases like this.

**Comment by alfred6427** [comment]
@ryrona2 I wonder if the issue is not related to the VPN kill switch.  It sounds like eBPF is used to filter/tag traffic, so I wonder if multicast traffic is not evaluated in the eBPF program, or maybe not assigned to the user profile.  

multicast range is 224.0.0.0/4, so one potential solution would be to block that range as part of the VPN kill switch.

**Comment by ryrona2** [comment]
@alfred6427 @thestinger There is something in the routing rules about tagged packets, so if either of you know where the eBPF code is, I would appreciate if you could give me a hint. I cannot promise I will get any time to look into this soon though, so anyone else is free to pick up the task. I don't know a think about eBPF either, even if I know iptables/nftables well, so it would probably be faster for someone else to find the issue anyway. Or maybe we should just report it upstream.

**Comment by alfred6427** [comment]
@ryrona2 I don't know much about eBPF or where the code would be.  I think they are typically .bp extension.
